### PR TITLE
Add SDK support for upcoming .NET Core 3.0 Managed COM Server support

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Add an MSBuild property and tasks/targets to enable creating a COM Server running on .NET Core.

If the user sets the `UseComHost` property and has an rid targeting Windows, the build will generate a `.clsidmap` file, make a copy of `comhost.dll`, embed the CLSIDmap if possible, and copy a renamed comhost.dll (along with the CLSIDmap if it wasn't embedded) to the output directory.

Right now since `comhost.dll` isn't in the AppHost package (it's still in PR review). I've disabled those tests for CI. To test them manually, add a dll in the `.nuget` folder in the apphost package next to `apphost.exe` and re-enable them.

Also, includes a property for internal use (i.e. those of us on the interop team) to explicitly not embed the CLSIDmap.

cc: @AaronRobinsonMSFT

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
This is the CLI side of the managed COM server support.

cc @vitek-karas @jeffschwMSFT @sbomer

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (AaronRobinsonMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: All source files should have a banner I believe.

File: src/Tasks/Microsoft.NET.Build.Tasks/ComHost.cs
Comment (AaronRobinsonMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/ComHost.cs]: Document origin of these values.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (AaronRobinsonMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: `.cslidmap`? Is this typo other places as well?

Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: It's just here.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateRegFreeComManifest.cs
Comment (jeffschwMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateRegFreeComManifest.cs]: What does Intermediate mean in this context?  Is this phrase exposed externally?

File: src/Tasks/Microsoft.NET.Build.Tasks/RegFreeComManifest.cs
Comment (jeffschwMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/RegFreeComManifest.cs]: Comments please

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAComServerLibrary.cs
Comment (jeffschwMSFT) [inline src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAComServerLibrary.cs]: With the recent decision to include runtimeconfig.json for these components, do these tests cover those scenarios as well?

**Review by jeffschwMSFT — CHANGES_REQUESTED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: That's one of the changes we discussed offline. We've decided to require users to explicitly specify a GUID for any type they want to expose via COM.

I can't think of any case for a user to have `[ComVisible(true)]` on a class and not specify an explicit GUID if the above statement still holds true. The other option (again assuming the statement above is still valid) would be to silently not expose the type to COM, which would result in a lot of unexpected behavior for users (and probably a lot of duplicate bug reports for us).

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I pulled this constant from the implementation of `System.Reflection.MetadataLoadContext` (see https://github.com/dotnet/corefx/blob/master/src/System.Reflection.MetadataLoadContext/src/System/Reflection/MetadataLoadContext.CoreAssembly.cs#L13).

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: MetadataReader returns a token to the direct base class of the type. Also, com objects are still valid if they inherit from other classes, even if those aren't marked ComVisible. See https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.comvisibleattribute?view=netframework-4.7.2#remarks.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateRegFreeComManifest.cs
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateRegFreeComManifest.cs]: "Intermediate" is the MSBuild term for the `obj/$Configuration/$TargetFramework/` directory. The `IntermediateAssembly` is the assembly directly output by `Csc` into the `obj` directory. It is later copied to the output directory. This terminology is used throughout MSBuild and the SDK.

File: src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAComServerLibrary.cs
Comment (jkoritzinsky) [reply src/Tests/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildAComServerLibrary.cs]: I haven't yet pushed out those changes.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (AaronRobinsonMSFT) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: >I can't think of any case for a user to have [ComVisible(true)] on a class and not specify an explicit GUID if the above statement still holds true.

That happens all the time and in fact is the normal case in .NET Framework. I think @jeffschwMSFT's point here is we want an explicit error or perhaps a wanring about what happened because this requirement will be new to users.

Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: This is giving an explicit error to the user. It will be (unless another PR takes the code before this is merged) error NETSDK1086. There error message is in Strings.resx. Do we want to downgrade this to a warning?

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Watch out here. this "flags" enum is a pit of failure.

The correct way to check visibility is to to mask with TypeAttributes.VisibilityMask (0x7) and see which of the eight values you get.

Above will be true for every odd visibility:  e.g. `(NestedPrivate=0x3) & (Public=0x1)` != 0)

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I see GetTypeName concats namespace . name so you can't use this directly. You should compare the namespace and name independently with `MetadataStringComparer`. 

You could make `bool HasTypeName(MetadataReader metadataReader  TypeReference baseClass,  string ns, string name)` and compare the namespace and names independently with `MetadataStringComparer`.

---
In reply to: [249175860](https://github.com/dotnet/sdk/pull/2814#discussion_r249175860) [](ancestors = 249175860)

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: This should not be plural.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I mean it should be `enum KnownType`

---
In reply to: [249180190](https://github.com/dotnet/sdk/pull/2814#discussion_r249180190) [](ancestors = 249180190)

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I think it would be better to throw BadImageFormatException here with a comment that there are no enum arguments to the attributes we care about. If ever we did encounter an enum argument to these attributes,  the stack will be more obvious if we fail now rather than reading the wrong number of bytes IMHO.

Side rant: The metadata format for attributes is really silly here in that it requires you to resolve an enum type to its definition to decide how many bytes to read. I wrote the custom attribute decoder and it pained me to have to have this method.

**Comment by nguerrera** [comment]
Sorry for the delay in reviewing so far. I still haven't had a chance to read the entire PR yet, so my comments aren't yet a complete review.

Also make note that we aren't branching for preview 2 so this will have to wait for us to be open for preview 3 before we can merge it.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: This is only considering the source-built binary. If we wanted to consider generating a comhost for a previously built binary, that would be done differently (and probably not as an MSBuild integration).

Comment (jeffschwMSFT) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: > This is only considering the source-built binary

This is good.  Then my only ask is to do a little more due diligence on the likelihood of a case where ComVisible is present, but was benign but is now an error.

Comment (jkoritzinsky) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: As a note, this task (and this error) will only execute on projects with the MSBuild `UseComHost` property set. Since this property currently doesn't exist, users will have to opt-in to this behavior.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: This should be `==`, not `!=`.

Consider switching over `type.Attributes & typeAttributes.VisibilityMask`

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: I'm not thrilled about yet another place where this is hard-coded. Hopefully, we won't invent another core library. I don't have a good solution for you.

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Or is this for json serialization to correct casing? Is there another mechanism for that?

File: src/Tasks/Microsoft.NET.Build.Tasks/RegFreeComManifest.cs
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/RegFreeComManifest.cs]: How big do we expect these clsidMap files to be? This should probably parsed via StreamReader -> JsonTextReader -> JObject.

**Comment by nguerrera** [comment]
This looks good overall. A few more small things after reading through everything. And we need to get the tests passing before we can merge. 

Separately, I would also like to discuss having the tasks for this live in a separate package. I've been discussing an approach with @fadimounir where we would have tasks provided from separate sdk(s) for runtime concerns like ready-to-run and single exe. We leave the targets that orchestrate these tasks and mutate build state in the SDK. 

I believe we can use the same approach for this. It would allow the code that is more coupled to runtime details to live closer to the runtime. I think it would make sense to move the app host creation/customization task there too.

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky The hosting libraries are all in. As soon as these changes are in and flow to the core-setup branch, we can update the core-setup authoring to include the comhost library in deployment and test the E2E in that repo.

**Comment by jkoritzinsky** [comment]
The unit tests in this PR are expecting the comhost library to be included in the apphost package. Can we add them to the package in core-setup? We'd merge this after updated packages with the comhost included have flowed over.

**Comment by AaronRobinsonMSFT** [comment]
Oops. Didn't realize that was a requirement for this insertion. Will do that.

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky Package has been updated - https://github.com/dotnet/core-setup/pull/5145

@nguerrera What would be the process for adding in some telemetry for when users set the `UseComHost` property to `true`?

**Comment by nguerrera** [comment]
I'll start a discussion on telemetry with the right folks. Let's handle that as a separate PR once we have a plan there.

**Comment by jkoritzinsky** [comment]
@nguerrera how do I update the version of the apphost/shared framework packages that the SDK pulls in to a version that has the comhost?

**Comment by nguerrera** [comment]
@jkoritzinsky First there needs to be a build of dotnet/core-sdk that includes it, then the version in global.json here needs to be updated to at least that. When did the apphost change go into core-setup. It might already be in from the automatic core-setup -> core-sdk insertions.

**Comment by nguerrera** [comment]
Looks like the core-setup -> core-sdk insertion is failing: https://github.com/dotnet/core-sdk/pull/474

Investigation should start there.

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky That insertion seems to have finally passed. Try now?

**Comment by jkoritzinsky** [comment]
@nguerrera there's some tests not related to the COM server support that fail when running with the new SDK. Any suggestions? Should I spin off those changes into a different PR?

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (jkotas) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: You just match the type name, and ignore the assembly name. If somebody defines their own `System.ValueType` or `System.Enum`, a lot of things will not work.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Good call.

**Comment by nguerrera** [comment]
@jkoritzinsky At a quick glance, it looks like CI was green on 312f555 with 3.0.100-preview3-010333 and then merge e5ac615 resolved a conflict in favor of a downgrade to 3.0.100-preview3-010324. 

Am I missing something? If not, I would expect going to 10333 or higher again should work.

**Comment by jkoritzinsky** [comment]
@nguerrera CI failed on 312f555 but due to some glitch, the check status never updated the PR (if you look at the checks that passed, you'll see that is was only the license/cla and WIP checks).

**Comment by nguerrera** [comment]
Ah! Ok, I'll look closer. :)

File: src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs
Comment (AaronRobinsonMSFT) [inline src/Tasks/Microsoft.NET.Build.Tasks/ClsidMap.cs]: Extra spaces. Line is good, but it looks like you have a tab/spaces.

**Comment by jkoritzinsky** [comment]
@nguerrera I finally found an SDK version that would build successfully on CI. Can you take another review pass or merge if you think it looks good?

**Review by nguerrera — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky Good job! When will we be able to consume this work in core-setup?

**Comment by jkoritzinsky** [comment]
As soon as we get an official build drop from the dotnet/core-sdk repo that has this in it, we can update core-setup to use that SDK version. Once that's updated, then we should be able to consume this in core-setup!

**Pull request merged**
